utils/pce-mkcd/pce-mkcd.cc

```diff
@@ -0,0 +1,765 @@
+/* pce-cd target .iso image builder.
+ *
+ * Copyright (c) 2023 Adrian Siekierka
+ *
+ * This file is part of the llvm-mos-sdk project and is redistributable under
+ * the terms of the Apache 2.0 license with the LLVM exceptions. See the LICENSE
+ * file in the project root for the full text.
+ */
+
+#include <algorithm>
+#include <cassert>
+#include <cctype>
+#include <cstdarg>
+#include <cstddef>
+#include <cstdint>
+#include <cstdlib>
+#include <cstring>
+#include <fstream>
+#include <ios>
+#include <map>
+#include <memory>
+#include <optional>
+#include <regex>
+#include <string>
+#include <vector>
+
+#include "../common/elf.h"
+#include "../common/elf-mos.h"
+
+#include "../common/parg/parg.h"
+
+// Program arguments, error handling.
+
+#define P_ARG_HELP 'h'
+#define P_ARG_QUIET 'q'
+#define P_ARG_VERBOSE 'v'
+#define P_ARG_ISO_OFFSET 128
+#define P_ARG_ISO_NO_PAD_END 129
+#define P_ARG_IPL 130
+
+static struct parg_option long_options[] = {
+  {"help", PARG_NOARG, 0, P_ARG_HELP},
+  {"ipl", PARG_REQARG, 0, P_ARG_IPL},
+  {"iso-no-pad-end", PARG_NOARG, 0, P_ARG_ISO_NO_PAD_END},
+  {"iso-offset", PARG_REQARG, 0, P_ARG_ISO_OFFSET},
+  {"quiet", PARG_NOARG, 0, P_ARG_QUIET},
+  {"verbose", PARG_NOARG, 0, P_ARG_VERBOSE},
+  {0, 0, 0, 0}
+};
+static const char *short_options = "hqv";
+
+static const char *long_option_descriptions[] = {
+  "Print help information.",
+  "Specify the path to the ipl.bin file (first CD sector).",
+  "Disable ISO padding mandated by the CD-ROM specification.",
+  "Offset at the beginning of the ISO, in sectors.",
+  "Disable progress messages.",
+  "Enable more verbose messages."
+};
+
+uint32_t iso_offset_sectors = 0;
+bool iso_pad = true;
+typedef enum {
+  VERBOSITY_QUIET = 0,
+  VERBOSITY_INFO = 1,
+  VERBOSITY_VERBOSE = 2
+} verbosity_level;
+verbosity_level verbosity = VERBOSITY_INFO;
+
+static void log(verbosity_level log_verbosity, const char *msg, ...) {
+  if (verbosity >= log_verbosity) {
+    va_list ap;
+    va_start(ap, msg);
+    vfprintf(stderr, msg, ap);
+    fprintf(stderr, "\n");
+  }
+}
+
+static void error(int result, const char *msg, ...) {
+  va_list ap;
+  va_start(ap, msg);
+  fprintf(stderr, "Error: ");
+  vfprintf(stderr, msg, ap);
+  fprintf(stderr, "\n");
+  exit(result);
+}
+
+// Vector utilities.
+
+static uint8_t byte_at(std::vector<char> vec, size_t offset) {
```

Comment:
const std::vector<char>&, here and everywhere; otherwise the standard requires this to memcpy the whole vector, which seems like it'll commonly be fairly large.

Reply:
Done! Thank you; I haven't worked with C++ anywhere near enough.

Reply:
Haha, or too much, depending on your perspective ;)